### PR TITLE
[Merged by Bors] - feat: more lemmas about the cocompact filter on topological spaces with a continuous multiplication and zero

### PR DESCRIPTION
**Pull request by trivial1711**
Let `M` be a topological space with a continuous multiplication operation and a zero. We prove the following lemmas.

- `tendsto_mul_nhds_zero_prod_of_disjoint_cocompact`: Let `l` be a filter on `M` which is disjoint from the cocompact filter. Then the multiplication map `M × M → M` tends to zero on the filter product `𝓝 0 ×ˢ l`. (We also prove a variant with the order of the product swapped.)
- `tendsto_mul_coprod_nhds_zero_inf_of_disjoint_cocompact`: Let `l` be a filter on `M × M` which is disjoint from the cocompact filter. Then the multiplication map `M × M → M` tends to zero on `Filter.coprod (𝓝 0) (𝓝 0) ⊓ l`.
- `tendsto_mul_nhds_zero_of_disjoint_cocompact`: Let `l` be a filter on `M × M` which is disjoint from the cocompact filter and less than or equal to `Filter.coprod (𝓝 0) (𝓝 0)`. Then the multiplication map `M × M → M` tends to zero on `l`.
- `Tendsto.tendsto_mul_zero_of_disjoint_cocompact`: Let `f : α → M` and `g : α → M` be functions. If `f` tends to zero on a filter `l` and the image of `l` under `g` is disjoint from the cocompact filter on `M`, then `fun (x : α) ↦ f x * g x` also tends to zero on `l`. (We also prove a variant with the order of the product swapped.)

These lemmas were suggested by a reviewer comment on #12313: https://github.com/leanprover-community/mathlib4/pull/12313#issuecomment-2124740887. These lemmas are intended to simplify the proof of `tendsto_mul_cocompact_nhds_zero` in #12313, which states that if `f : α → M` and `g : β → M` are continuous functions that both tend to zero on the cocompact filter, then the "outer product" `fun (i : α × β) ↦ (f i.1) * (g i.2)` also tends to zero on the cocompact filter.

---

- [x] depends on: #19718
- [x] depends on: #19717
- [x] depends on: #19716 
[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Algebra/Monoid.lean [hunk 1]
```diff
@@ -399,6 +399,103 @@ instance Submonoid.continuousMul [TopologicalSpace M] [Monoid M] [ContinuousMul
...
+variable {α β : Type*}
+variable [TopologicalSpace M] [MulZeroClass M] [ContinuousMul M]
+
+/-- Let `M` be a topological space with a continuous multiplication operation and a `0`.
+Let `l` be a filter on `M` which is disjoint from the cocompact filter. Then, the multiplication map
+`M × M → M` tends to zero on the filter product `𝓝 0 ×ˢ l`. -/
+theorem tendsto_mul_nhds_zero_prod_of_disjoint_cocompact {l : Filter M}
+    (hl : Disjoint l (cocompact M)) :
+    Tendsto (fun (x : M × M) ↦ x.1 * x.2) (𝓝 0 ×ˢ l) (𝓝 0) := calc
+  (𝓝 0 ×ˢ l).map fun (x : M × M) ↦ x.1 * x.2
```
Comment (ADedecker) [inline Mathlib/Topology/Algebra/Monoid.lean:415]: We tend to prefer `map f l` over `l.map f` for consistency with the `f '' s` notation, where `f` comes first.

File: Mathlib/Topology/Algebra/Monoid.lean [hunk 2]
```diff
@@ -399,6 +399,103 @@ instance Submonoid.continuousMul [TopologicalSpace M] [Monoid M] [ContinuousMul
...
+  _ ≤ (𝓝ˢ ({0} ×ˢ Set.univ)).map fun (x : M × M) ↦ x.1 * x.2 :=
+    Filter.map_mono <| nhds_prod_le_of_disjoint_cocompact 0 hl
```
Comment (ADedecker) [inline Mathlib/Topology/Algebra/Monoid.lean:417]:
```suggestion
    map_mono <| nhds_prod_le_of_disjoint_cocompact 0 hl
```

File: Mathlib/Topology/Algebra/Monoid.lean [hunk 3]
```diff
@@ -399,6 +399,103 @@ instance Submonoid.continuousMul [TopologicalSpace M] [Monoid M] [ContinuousMul
...
+  _ ≤ 𝓝 0 := by
+    apply continuous_mul.tendsto_nhdsSet_nhds
+    rintro x ⟨_, _⟩
+    simp_all
```
Comment (ADedecker) [inline Mathlib/Topology/Algebra/Monoid.lean:421]:
```suggestion
  _ ≤ 𝓝 0 := continuous_mul.tendsto_nhdsSet_nhds fun _ ⟨hx, _⟩ ↦ mul_eq_zero_of_left hx _
```

File: Mathlib/Topology/Algebra/Monoid.lean [hunk 4]
```diff
@@ -399,6 +399,103 @@ instance Submonoid.continuousMul [TopologicalSpace M] [Monoid M] [ContinuousMul
...
+
+/-- Let `M` be a topological space with a continuous multiplication operation and a `0`.
+Let `l` be a filter on `M × M` which is disjoint from the cocompact filter. Then, the multiplication
+map `M × M → M` tends to zero on `Filter.coprod (𝓝 0) (𝓝 0) ⊓ l`. -/
+theorem tendsto_mul_coprod_nhds_zero_inf_of_disjoint_cocompact {l : Filter (M × M)}
+    (hl : Disjoint l (cocompact (M × M))) :
+    Tendsto (fun (x : M × M) ↦ x.1 * x.2) (Filter.coprod (𝓝 0) (𝓝 0) ⊓ l) (𝓝 0) := by
```
Comment (ADedecker) [inline Mathlib/Topology/Algebra/Monoid.lean:442]:
```suggestion
    Tendsto (fun (x : M × M) ↦ x.1 * x.2) ((𝓝 0).coprod (𝓝 0) ⊓ l) (𝓝 0) := by
```

File: Mathlib/Topology/Algebra/Monoid.lean [hunk 5]
```diff
@@ -399,6 +399,103 @@ instance Submonoid.continuousMul [TopologicalSpace M] [Monoid M] [ContinuousMul
...
+  have := calc
+    Filter.coprod (𝓝 0) (𝓝 0) ⊓ l
+    _ ≤ Filter.coprod (𝓝 0) (𝓝 0) ⊓ l.map Prod.fst ×ˢ l.map Prod.snd      :=
+      inf_le_inf_left _ le_prod_map_fst_snd
+    _ ≤ 𝓝 0 ×ˢ l.map Prod.snd ⊔ l.map Prod.fst ×ˢ 𝓝 0                    :=
+      Filter.coprod_inf_prod_le _ _ _ _
+  apply Tendsto.mono_left (h := this)
+  apply Tendsto.sup
```
Comment (ADedecker) [inline Mathlib/Topology/Algebra/Monoid.lean:450]:
```suggestion
  apply (Tendsto.sup _ _).mono_left this
```

File: Mathlib/Topology/Algebra/Monoid.lean [hunk 6]
```diff
@@ -399,6 +399,103 @@ instance Submonoid.continuousMul [TopologicalSpace M] [Monoid M] [ContinuousMul
...
+This is the same theorem as `Tendsto.tendsto_mul_zero_of_disjoint_cocompact'`, except the
+hypotheses on `f` and `g` are swapped. -/
+theorem Tendsto.tendsto_mul_zero_of_disjoint_cocompact {f g : α → M} {l : Filter α}
+    (hf : Tendsto f l (𝓝 0)) (hg : Disjoint (map g l) (cocompact M)) :
+    Tendsto (fun x ↦ f x * g x) l (𝓝 0) := calc
+  l.map (fun x ↦ f x * g x)
+  _ = (l.map (fun x ↦ (f x, g x))).map (fun (x : M × M) ↦ x.1 * x.2) := by rw [map_map]; rfl
+  _ ≤ (𝓝 0 ×ˢ map g l).map (fun (x : M × M) ↦ x.1 * x.2) :=
+    map_mono (hf.prod_mk Filter.tendsto_map)
+  _ ≤ 𝓝 0 := tendsto_mul_nhds_zero_prod_of_disjoint_cocompact hg
```
Comment (ADedecker) [inline Mathlib/Topology/Algebra/Monoid.lean:479]:
```suggestion
    Tendsto (fun x ↦ f x * g x) l (𝓝 0) := 
  tendsto_mul_nhds_zero_prod_of_disjoint_cocompact hg |>.comp (hf.prod_mk tendsto_map)
```

File: Mathlib/Topology/Algebra/Monoid.lean [hunk 7]
```diff
@@ -399,6 +399,103 @@ instance Submonoid.continuousMul [TopologicalSpace M] [Monoid M] [ContinuousMul
...
+  simpa [inf_eq_right.mpr h'l] using tendsto_mul_coprod_nhds_zero_inf_of_disjoint_cocompact hl
+
+/-- Let `M` be a topological space with a continuous multiplication operation and a `0`.
+Let `f : α → M` and `g : α → M` be functions. If `f` tends to zero on a filter `l`
+and the image of `l` under `g` is disjoint from the cocompact filter on `M`, then
+`fun (x : α) ↦ f x * g x` also tends to zero on `l`.
+
+This is the same theorem as `Tendsto.tendsto_mul_zero_of_disjoint_cocompact'`, except the
+hypotheses on `f` and `g` are swapped. -/
+theorem Tendsto.tendsto_mul_zero_of_disjoint_cocompact {f g : α → M} {l : Filter α}
```
Comment (ADedecker) [inline Mathlib/Topology/Algebra/Monoid.lean:472]: What do you think about naming this one `Tendsto.mul_zero_of_disjoint_cocompact_right` and the other one `Tendsto.mul_zero_of_disjoint_cocompact_left`?

**Comment by trivial1711** [comment]
Thanks for the comments. I have made the edits.

File: Mathlib/Topology/Algebra/Monoid.lean [hunk 8]
```diff
@@ -399,6 +399,81 @@ instance Submonoid.continuousMul [TopologicalSpace M] [Monoid M] [ContinuousMul
...
+Let `l` be a filter on `M × M` which is disjoint from the cocompact filter. Then, the multiplication
+map `M × M → M` tends to zero on `(𝓝 0).coprod (𝓝 0) ⊓ l`. -/
+theorem tendsto_mul_coprod_nhds_zero_inf_of_disjoint_cocompact {l : Filter (M × M)}
+    (hl : Disjoint l (cocompact (M × M))) :
+    Tendsto (fun (x : M × M) ↦ x.1 * x.2) ((𝓝 0).coprod (𝓝 0) ⊓ l) (𝓝 0) := by
+  have := calc
+    (𝓝 0).coprod (𝓝 0) ⊓ l
+    _ ≤ (𝓝 0).coprod (𝓝 0) ⊓ map Prod.fst l ×ˢ map Prod.snd l            :=
+      inf_le_inf_left _ le_prod_map_fst_snd
+    _ ≤ 𝓝 0 ×ˢ map Prod.snd l ⊔ map Prod.fst l ×ˢ 𝓝 0                    :=
```
Comment (jcommelin) [inline Mathlib/Topology/Algebra/Monoid.lean:441]:
```suggestion
    _ ≤ (𝓝 0).coprod (𝓝 0) ⊓ map Prod.fst l ×ˢ map Prod.snd l :=
      inf_le_inf_left _ le_prod_map_fst_snd
    _ ≤ 𝓝 0 ×ˢ map Prod.snd l ⊔ map Prod.fst l ×ˢ 𝓝 0 :=
```

**Review by jcommelin — COMMENTED** [review]
bors d+

**Pull request closed without merging**